addons/controllers/utils/utils.go

```diff
@@ -134,3 +135,57 @@ func TryParseClusterVariableString(ctx context.Context, cluster *clusterapiv1bet
 	}
 	return res
 }
+
+// GetOwnerCluster returns cluster owner for the object.
+// Returns cluster=nil and err if no cluster can be determined as owner.
+// Returns cluster object with only name and namespace  and err,
+// if details for the cluster determined to be an woner can not be found.
+func GetOwnerCluster(ctx context.Context, k8sClient client.Client, k8sObject client.Object, namespace, pkgRefName string) (*clusterapiv1beta1.Cluster, error) {
+	logger := log.FromContext(ctx)
+	cluster := &clusterapiv1beta1.Cluster{}
+	clusterName := ""
+
+	for _, ownerRef := range k8sObject.GetOwnerReferences() {
+		if strings.EqualFold(ownerRef.Kind, constants.ClusterKind) {
+			clusterName = ownerRef.Name
+			break
+		}
+	}
+	if clusterName != "" {
+		cluster.Name = clusterName
+		cluster.Namespace = namespace
+	} else {
+		// Means no cluster listed
+		// look at object name and match against patern of {clustern-name}-{packagerefname}-package
+		possibleClusterName := util.GeneratePackageClusterName(k8sObject.GetName(), pkgRefName)
+		if possibleClusterName != "" {
+			cluster.Name = possibleClusterName
+			cluster.Namespace = k8sObject.GetNamespace()
+		} else {
+			errorString := "no owner cluster could be determined"
+			logger.Info(errorString)
+			err := errors.New(errorString)
+			return nil, err
+		}
+	}
```

Comment:
Do we also need to match against an existing clustername?  it could expand the options. 
the case it would handle is one where, cluster does not have owner refs, does not have a name matching pattern.... 

I think it might cause problems in the case it tries to reconcile a custom addons-config which has been genrated before the matching custom clusterbootstrap has been created.  
It could end up being a bug because if by coincidence a cluster with matching name exists, it would incorrectly assuem the new config belongs to it.

Reply:
we only need to match on owner reference